src/common/HipGridScan.hpp

```diff
@@ -101,99 +104,135 @@ struct GridScan
+      ballot_type prev_blocks_ready_ballot = ballot_type(0);
+      ballot_type prev_grids_ready_ballot = ballot_type(0);
 
-          prev_block_base_id -= warp_size;
-        }
+      // accumulate full warp worths of block counts
+      // stop if run out of full warps of a grid count is ready
+      while (prev_block_base_id >= 0) {
 
         const int prev_block_id = prev_block_base_id + warp_index;
 
         // ensure previous block_counts are ready
-        // this checks that block counts is ready for all blocks above
-        // the highest grid count that is ready
-        while (~prev_blocks_ready_ballot >= prev_grids_ready_ballot) {
-
-          if (prev_block_id >= 0) {
-            prev_block_ready = atomicCAS(&block_readys[prev_block_id], 11u, 11u);
-          }
+        do {
+          prev_block_ready = atomicCAS(&block_readys[prev_block_id], 11u, 11u);
 
           prev_blocks_ready_ballot = __ballot(prev_block_ready >= 1u);
-          prev_grids_ready_ballot = __ballot(prev_block_ready == 2u);
+
+        } while (prev_blocks_ready_ballot != ~ballot_type(0));
+
+        prev_grids_ready_ballot = __ballot(prev_block_ready == 2u);
+
+        if (prev_grids_ready_ballot != ballot_type(0)) {
+          break;
         }
+
         __threadfence(); // ensure block_counts or grid_counts ready (acquire)
 
-        // read one grid_count from a block with id grid_count_ready_id
-        // and read the block_counts from blocks with higher ids.
-        if (warp_index_mask > prev_grids_ready_ballot) {
-          // accumulate block_counts for prev_block_id
-          prev_grid_count += block_counts[prev_block_id];
-        } else if (prev_grids_ready_ballot == (prev_grids_ready_ballot & warp_index_mask_right)) {
-          // accumulate grid_count for grid_count_ready_id
-          prev_grid_count += grid_counts[prev_block_id];
-        }
+        // accumulate block_counts for prev_block_id
+        prev_grid_count += block_counts[prev_block_id];
 
+        prev_block_ready = 0u;
 
-        WarpReduce().reduce(prev_grid_count, prev_grid_count, s_temp_storage.warp_reduce_storage);
-        prev_grid_count = __shfl(prev_grid_count, 0, warp_size); // broadcast output to all threads in warp
+        prev_block_base_id -= warp_size;
+      }
 
-        if (last_thread) {
+      const int prev_block_id = prev_block_base_id + warp_index;
 
-          if (!last_block) {
-            grid_counts[block_id] = prev_grid_count + inclusive[items_per_thread-1];   // write inclusive scan result for grid through block
-            __threadfence();                        // ensure grid_counts ready (release)
-            atomicExch(&block_readys[block_id], 2u); // write grid_counts is ready
-          }
+      // ensure previous block_counts are ready
+      // this checks that block counts is ready for all blocks above
+      // the highest grid count that is ready
+      while (~prev_blocks_ready_ballot >= prev_grids_ready_ballot) {
```

Comment:
This was always false leading to a hang. The most significant bits of the numbers returned from ballot were always 0 when the warp size was 32 but the numbers were 64-bit.